packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -837,6 +854,224 @@ function rerenderReducer<S, I, A>(
+      const handleChange = () => {
+        const latestGetSnapshot = memoizedState.getSnapshot;
+        try {
+          setSnapshot(latestGetSnapshot(source._source));
+
+          // Record a pending mutable source update with the same expiration time.
+          const expirationTime = (stateHook.queue: any).pending.expirationTime;
+          setPendingExpirationTime(root, expirationTime);
+        } catch (error) {
+          // A selector might throw after a source mutation.
+          // e.g. it might try to read from a part of the store that no longer exists.
+          // In this case we should still schedule an update with React.
+          // Worst case the selector will throw again and then an error boundary will handle it.
+          setSnapshot(() => {
+            throw error;
+          });
+        }
+      };
+
+      const unsubscribe = subscribe(source._source, handleChange);
+      invariant(
+        typeof unsubscribe === 'function',
+        'Mutable source subscribe function must return an unsubscribe function.',
+      );
+
+      // Check for a possible change between when we last rendered and when we just subscribed.
+      const maybeNewVersion = getVersion();
+      if (version !== maybeNewVersion) {
+        const maybeNewSnapshot = getSnapshot(source._source);
+        if (snapshot !== maybeNewSnapshot) {
+          setSnapshot(maybeNewSnapshot);
+        }
+      }
+
+      return unsubscribe;
+    }
+  }, [source, subscribe, getSnapshot]);
+
+  if (prevSource !== source || prevSubscribe !== subscribe) {
+    // If either the source or the subscription have changed,
+    // we need to treat the read as potentially unsafe,
+    // since we can't rely on the current update queue.
+    snapshot = readFromUnsubcribedMutableSource(source, getSnapshot);
+
+    // Throw away any pending updates since they are no longer relevant.
+    const queue = (stateHook.queue = {
+      ...stateHook.queue,
+      pending: null,
+      lastRenderedState: snapshot,
+    });
+    setSnapshot = queue.dispatch = (dispatchAction.bind(
+      null,
+      currentlyRenderingFiber,
+      queue,
+    ): any);
+    hook.memoizedState = hook.baseState = snapshot;
+  } else if (prevGetSnapshot !== getSnapshot) {
+    const maybeNewSnapshot = getSnapshot(source._source);
+
+    // Normally a new getSnapshot function implies a new snapshot value.
+    // This function should be memoized so that it only changes when its dependencies change.
+    // Tf the returned snapshot value is the same-
+    // it likely means getSnapshot is not memoized correctly (e.g. inline function).
+    // In this case we can avoid the expensive deopt and just update local component state.
+    if (snapshot !== maybeNewSnapshot) {
+      // Even if the snapshot value did change, it might be okay to read the new value.
+      // If it's not we'll throw and restart the render (to avoid tearing).
+      snapshot = readFromUnsubcribedMutableSource(source, getSnapshot);
```

Comment:
FYI: I will want to fork `readFromUnsubcribedMutableSource` rather than calling it again (just in case snapshot calculation is expensive; there's no need to re-run that function) but for now I'm going to hold off on this change until I get feedback on the overall approach.